Mathlib/Data/List/TakeDrop.lean

```diff
@@ -150,6 +150,69 @@ theorem span_eq_takeWhile_dropWhile (l : List α) : span p l = (takeWhile p l, d
 
 end Filter
 
+section Splits
+
+/-! ### Generate splits of a list. -/
+
+/-- Pairs of prefixes and suffixes of a list. -/
+def splits (xs : List α) : List (List α × List α) :=
+  List.map xs.splitAt (range (xs.length + 1))
+
+@[simp]
+lemma splits_nil : (List.nil (α:=α)).splits= [([], [])] := by rfl
+
+@[simp]
+lemma splits_cons (x : α) (xs : List α) :
+    List.splits (x :: xs) =
+    ([], x :: xs) :: List.map (fun td ↦ (x :: td.1, td.2)) (List.splits xs) := by
+  simp [List.splits, List.range_succ_eq_map]
+
+lemma splits_spec (x y l : List α) :
+    (x, y) ∈ l.splits ↔ l = x ++ y := by
+  simp [List.splits]
+  constructor
+  · rintro ⟨n, hn, htake, hdrop⟩
+    rw [←htake, ←hdrop, List.take_append_drop]
+  · rintro rfl
+    exists x.length
+    constructor
+    · simp; omega
+    · constructor
+      · rw [List.take_left]
+      · rw [List.drop_left]
+
+lemma splits_non_empty (l : List α) : ∃ td, td ∈ l.splits := by
```

Comment:
I would write this lemma with explicit choice for the member, i.e
```
lemma nil_self_mem_splits (l : List α) : ([], l) ∈ l.splits := by
```
or
```
lemma self_nil_mem_splits (l : List α) : (l, []) ∈ l.splits := by
```
or better yet, both.